File: editors/chroma/gengo.go
// Package chroma contains a Gengoscript lexer for Chroma (used by Hugo and other Go-based syntax highlighters).
package chroma

import (
	"github.com/alecthomas/chroma/v2"
)

// Gengoscript is a Chroma lexer for the Gengoscript programming language.
var Gengoscript = chroma.MustNewLexer(
	&chroma.Config{
		Name:      "Gengoscript",
		Aliases:   []string{"gengo"},
		Filenames: []string{"*.gengo"},
	},
	func() chroma.Rules {
		return chroma.Rules{
			"root": {
				// Comments
				{`/\*`, chroma.CommentMultiline, chroma.Push("block_comment")},
				{`//.*`, chroma.CommentSingle, nil},

				// Keywords
				{`\b(assert|break|case|const|continue|cycle|default|defer|else|enum|false|for|func|if|import|in|interface|null|predicate|pub|range|return|struct|subtype|switch|test|trap|true|type|var|variant)\b`, chroma.Keyword, nil},

				// Built-in types
				{`\b(int|float|bool|string|any)\b`, chroma.KeywordType, nil},

				// Numbers: decimal int, float, scientific
				{`\b-?[0-9][0-9_]*(?:\.[0-9][0-9_]*)?(?:[eE][+-]?[0-9]+)?\b`, chroma.Number, nil},

				// Double-quoted strings with escapes
				{`"(?:[^"\\]|\\.)*"`, chroma.StringDouble, nil},

				// Single-quoted raw strings
				{`'[^']*'`, chroma.StringSingle, nil},

				// Multiline raw string lines
				{`^\\\\`, chroma.StringOther, chroma.Push("multiline")},

				// Rune literals
				{"`[^`]`", chroma.StringChar, nil},

				// Whitespace
				{`\s+`, chroma.TextWhitespace, nil},
			},
			"block_comment": {
				{`\*/`, chroma.CommentMultiline, chroma.Pop(1)},
				{`[^*]+`, chroma.CommentMultiline, nil},
				{`\*`, chroma.CommentMultiline, nil},
			},
			"multiline": {
				{`$`, chroma.StringOther, chroma.Pop(1)},
				{`.*$`, chroma.StringOther, nil},
			},
		}
	},
)

File: editors/highlight.js/gengo.js
/*
 * highlight.js grammar for Gengo.
 * Usage: hljs.registerLanguage('gengo', gengo);
 */
function gengo(hljs) {
  var KEYWORDS = {
    keyword:
      'assert break case const continue cycle default defer else enum ' +
      'false for func if import in interface null predicate pub range ' +
      'return struct subtype switch test trap true type var variant',
    type: 'int float bool string any',
  };

  var ESCAPE = {
    className: 'attr',
    begin: /\\(?:[ntr\\"'/]|x[0-9a-fA-F]{2})/,
  };

  var STRING_DQ = {
    className: 'string',
    begin: '"',
    end: '"',
    contains: [ESCAPE],
  };

  var STRING_SQ = {
    className: 'string',
    begin: "'",
    end: "'",
  };

  var MULTILINE = {
    className: 'string',
    begin: /^\\\\/m,
    end: /$/m,
  };

  var RUNE = {
    className: 'string',
    match: /`[^`]`/,
  };

  var NUMBER = {
    className: 'number',
    relevance: 0,
    variants: [
      { begin: /\b-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?\b/ },
    ],
  };

  var COMMENT = {
    className: 'comment',
    begin: /\/\//,
    end: /$/,
  };

  return {
    name: 'Gengo',
    aliases: ['gengo'],
    keywords: KEYWORDS,
    contains: [
      COMMENT,
      STRING_DQ,
      STRING_SQ,
      MULTILINE,
      RUNE,
      NUMBER,
    ],
  };
}

module.exports = gengo;
